src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,514 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace System.Collections.Generic
+{
+    /// <summary>
+    /// Represents a data structure in which each element additionally has a priority
+    /// associated with it.
+    /// </summary>
+    /// <typeparam name="TElement">The type of the element.</typeparam>
+    /// <typeparam name="TPriority">The type of the priority.</typeparam>
+    public class PriorityQueue<TElement, TPriority>
+    {
+        /// <summary>
+        /// Represents an implicit heap-ordered complete d-ary tree, stored as an array.
+        /// </summary>
+        private readonly List<(TElement element, TPriority priority)> nodes;
+
+        private const int RootIndex = 0;
```

Comment:
Why is this needed?

Reply:
I intended to avoid magic values in the code (represent them with constants instead) and make the code readable/verbose/clear even to someone who is very tired and drunk 😄  Is there a disadvantage to this approach?

Reply:
I think it's pretty obvious to most programmers that `0` represents the first index. If anything it makes reading the code harder since I now need to look up what `RootIndex` is.

Reply:
I'm not merely referring to the first index. We work with a quaternary tree that is implicitly represented in an array, the constant refers to the index in the underlying array that contains the root node of that tree. Far from obvious when all you have to work with is a number vs a named constant that tells you it's the root.

Reply:
In any array heap the first index points to the root node if it exists. Not sure why this needs to be lifted to a constant.